core/src/main/scala/kafka/log/Log.scala

```diff
@@ -169,6 +169,12 @@ class Log(@volatile var dir: File,
    */
   @volatile var firstUnstableOffset: Option[LogOffsetMetadata] = None
 
+  /* Keep track of the current high watermark in order to ensure that segments containing offsets at or above it are
+   * not eligible for deletion. This is needed to prevent the log start offset (which is exposed in fetch responses)
+   * from getting ahead of the high watermark.
+   */
+  @volatile private var replicaHighWatermark: Option[Long] = None
```

Comment:
Yes, exactly. I considered that, but decided to use the door that was already open. Either way, conceptually, the layering has been broken since the log needs to be aware of the high watermark. I'm not sure there's a way around this, so I will be taking a hard look at merging `Replica` and `Log` into a single object in a follow-up.